src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp

```diff
@@ -37,4 +37,44 @@ void three_index_constraint(
     const tnsr::iaa<DataType, SpatialDim, Frame>& d_spacetime_metric,
     const tnsr::iaa<DataType, SpatialDim, Frame>& phi) noexcept;
 // @}
+
+// @{
+/*!
+ * \brief Computes the generalized-harmonic gauge constraint.
+ *
+ * \details Computes the generalized-harmonic gauge constraint
+ * [Eq. (40) of http://arXiv.org/abs/gr-qc/0512093v3],
+ * \f[
+ * C_a = H_a + g^{ij} \Phi_{ija} + t^b \Pi_{ba}
+ * - \frac{1}{2} g^i_a \psi^{bc} \Phi_{ibc}
+ * - \frac{1}{2} t_a \psi^{bc} \Pi_{bc},
+ * \f]
+ * where \f$H_a\f$ is the gauge function,
+ * \f$\psi_{ab}\f$ is the spacetime metric,
+ * \f$\Pi_{ab}=-t^c\partial_c \psi_{ab}\f$, and
+ * \f$\Phi_{iab} = \partial_i\psi_{ab}\f$; \f$t^a\f$ is the timelike unit
+ * normal one form to the spatial slice, \f$g^{ij}\f$ is the inverse spatial
+ * metric, and \f$g^b_c = \delta^b_c + t^b t_c\f$.
+ */
+template <size_t SpatialDim, typename Frame, typename DataType>
+tnsr::a<DataType, SpatialDim, Frame> gauge_constraint(
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_function,
+    const tnsr::a<DataType, SpatialDim, Frame>& spacetime_normal_one_form,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_normal_vector,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::AA<DataType, SpatialDim, Frame>& inverse_spacetime_metric,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi) noexcept;
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void gauge_constraint(
```

Comment:
Is this version of the function tested?

Reply:
Unless I screwed something up. The check_with_random_values test calls this one (return by value), which then calls the other one (return by pointer), thus testing both. Or at a least, that's what's supposed to happen...did I screw up somewhere?

Reply:
I'd like to see this function explicitly tested though, since the fact that one calls the other is an implementation detail. Since you've two tests, the analytic and the random version, I'd be fine with just adding a call to this version into the analytic test, and checking that it also gives zero.